tests/commission_server/border_agent_session.cpp

```diff
@@ -0,0 +1,379 @@
+        aCoapAgent->FreeMessage(message);
+
+        do
+        {
+            ret = mbedtls_ssl_read(&mSsl, buffer, sizeof(buffer));
+            if (ret > 0)
+            {
+                aCoapAgent->Input(buffer, static_cast<uint16_t>(ret), NULL, 0);
+                switch (mCommissionState)
+                {
+                case kStateConnected:
+                    ret = MBEDTLS_ERR_SSL_WANT_READ;
+                default:
+                    break;
+                }
+            }
+        } while (ret == MBEDTLS_ERR_SSL_WANT_READ || ret == MBEDTLS_ERR_SSL_WANT_WRITE);
+    }
+
+    otbrLog(OTBR_LOG_INFO, "COMM_PET.req: complete");
+
+    ret = mCommissionState == kStateAccepted ? 0 : -1;
+    return ret;
+}
+
+/** handle c/cp response */
+void BorderAgentDtlsSession::HandleCommissionerPetition(const Coap::Message &aMessage, void *aContext)
+{
+    uint16_t                length;
+    int                     tlvType;
+    const Tlv *             tlv;
+    const uint8_t *         payload;
+    BorderAgentDtlsSession *session = reinterpret_cast<BorderAgentDtlsSession *>(aContext);
+
+    otbrLog(OTBR_LOG_INFO, "COMM_PET.rsp: start");
+    payload = aMessage.GetPayload(length);
+    tlv     = reinterpret_cast<const Tlv *>(payload);
+
+    while (utils::LengthOf(payload, tlv) < length)
+    {
+        int8_t state = static_cast<int8_t>(tlv->GetValueUInt8());
+
+        tlvType = tlv->GetType();
+        switch (tlvType)
+        {
+        case Meshcop::kState:
+            switch (state)
+            {
+            case kStateAccepted:
+                otbrLog(OTBR_LOG_INFO, "COMM_PET.rsp: state=accepted");
+                session->mCommissionState = kStateAccepted;
+                break;
+            case kStateRejected:
+                otbrLog(OTBR_LOG_INFO, "COMM_PET.rsp: state=accepted");
+                session->mCommissionState = kStateRejected;
+                break;
+            default:
+                otbrLog(OTBR_LOG_INFO, "COMM_PET.rsp: state=%d", state);
+                session->mCommissionState = kStateInvalid;
+                break;
+            }
+            break;
+        case Meshcop::kCommissionerSessionId:
+            session->mCommissionerSessionId = tlv->GetValueUInt16();
+            otbrLog(OTBR_LOG_INFO, "COMM_PET.rsp: session-id=%d", session->mCommissionerSessionId);
+            break;
+
+        default:
+            otbrLog(OTBR_LOG_INFO, "COMM_PET.rsp: ignore-tlv: %d", tlvType);
+            break;
+        }
+        tlv = tlv->GetNext();
+    }
+
+    otbrLog(OTBR_LOG_INFO, "COMM_PET.rsp: complete");
+}
+
+int BorderAgentDtlsSession::SetupProxyServer()
+{
+    int         ret;
+    int         optval = 1;
+    sockaddr_in addr;
+    mListenFd            = -1;
+    addr.sin_family      = AF_INET;
+    addr.sin_port        = htons(FORWARD_PORT);
+    addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    mListenFd            = socket(AF_INET, SOCK_STREAM, 0);
```

Comment:
Still have to choose TCP to do the proxy since it allows easy connection management. I can easily know when a client disconnects and we no longer relay message from NCP to it. Such message relaying is now broadcast to every client, we don't have a built-in ending mechanism.